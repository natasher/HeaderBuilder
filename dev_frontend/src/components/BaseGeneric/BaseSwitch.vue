<template>
  <div :class="classObject">

    <div class="switch" :data-name="fieldName">

      <span class = "path"   :data-name = "fieldName"></span>
      <span class = "handle" :data-name = "fieldName"></span>

    </div>
    <label>{{ label }}</label>

  </div>
</template>

<script>
export default {

  name: 'BaseSwitch',

  props: {
    /**
     * Label shown next to switch.
     */
    label: {
      type    : String,
      required: true,
    },
    /**
     * Determinates if switch is active,
     * takes function that returns `boolean` value
     * or boolean itself.
     */
    isActive: {
      type    : [ Function, Boolean ],
      required: false,
      default : false,
    },
    /**
     * The name of the containing/binded field.
     */
    fieldName: {
      type    : [ String, Boolean ],
      required: false,
      default : false,
    }
  },

  computed: {
    classObject: function () {
      return {
        'mfn-field mfnf-checkbox': true,
        'active'                 : this.isActive
      }
    }
  },

}
</script>
