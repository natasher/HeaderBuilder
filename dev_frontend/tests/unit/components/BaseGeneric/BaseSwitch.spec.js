import { shallowMount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import merge from 'lodash.merge'
import BaseSwitch from '../../../../src/components/BaseGeneric/BaseSwitch.vue'

const localVue = createLocalVue()

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    propsData: {
      label: 'test switch'
    }
  }

  return shallowMount( BaseSwitch, merge( defaultMountingOptions, overrides ))
}

describe('BaseSwitch.vue', () => {

  describe('snapshots:', () => {

    test('should match snapshot if active', async () => {
      const wrapper = createWrapper({
        propsData: {
          isActive: true,
          fieldName: 'testField'
        }
      })
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if inactive', async () => {
      const wrapper = createWrapper({
        propsData: {
          fieldName: 'testField'
        }
      })
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  test('should render label from props', async () => {
    const label = 'my awesome label'
    const wrapper = createWrapper({
      propsData: {
        label,
      }
    })
    await flushPromises()

    expect( wrapper.find( 'label' ).text() ).toBe( label )
  })

  test('should contain field and field checkbox classes if inactive', async () => {
    const wrapper = createWrapper()
    await flushPromises()

    expect( wrapper.classes().length ).toBe( 2 )
  })

  test('should contain field and field checkbox classes and .active if active', async () => {
    const wrapper = createWrapper({
      propsData: {
        isActive: true
      }
    })
    await flushPromises()

    expect( wrapper.classes().length ).toBe( 3 )
    expect( wrapper.classes() ).toContain( 'active' )
  })

  test('should pass data-name attribute to switch elements', async () => {
    const fieldName = 'awesomeSwitch'
    const wrapper = createWrapper({
      propsData: {
        fieldName,
      }
    })
    await flushPromises()

    expect( wrapper.find('.switch').attributes()[ 'data-name' ]).toBe( fieldName )
    expect( wrapper.find('.path').attributes()[ 'data-name' ]).toBe( fieldName )
    expect( wrapper.find('.handle').attributes()[ 'data-name' ]).toBe( fieldName )
  })

})