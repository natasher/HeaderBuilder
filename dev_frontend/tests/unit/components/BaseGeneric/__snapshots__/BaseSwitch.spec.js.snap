// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSwitch.vue snapshots: should match snapshot if active 1`] = `
<div
  class="mfn-field mfnf-checkbox active"
>
  <div
    class="switch"
    data-name="testField"
  >
    <span
      class="path"
      data-name="testField"
    />
     
    <span
      class="handle"
      data-name="testField"
    />
  </div>
   
  <label>
    test switch
  </label>
</div>
`;

exports[`BaseSwitch.vue snapshots: should match snapshot if inactive 1`] = `
<div
  class="mfn-field mfnf-checkbox"
>
  <div
    class="switch"
    data-name="testField"
  >
    <span
      class="path"
      data-name="testField"
    />
     
    <span
      class="handle"
      data-name="testField"
    />
  </div>
   
  <label>
    test switch
  </label>
</div>
`;
